userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "logger.h"
+
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <list>
+#include <string>
+
+namespace sysdig
+{
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::async_metadata_source(
+		const uint64_t max_wait_ms,
+		const uint64_t ttl_ms):
+	m_max_wait_ms(max_wait_ms),
+	m_ttl_ms(ttl_ms),
+	m_thread(),
+	m_running(false),
+	m_terminate(false),
+	m_mutex(),
+	m_queue_not_empty_condition(),
+	m_metadata_map()
+{ }
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::~async_metadata_source()
+{
+	try
+	{
+		stop();
+	}
+	catch(...)
+	{
+		g_logger.log(std::string(__FUNCTION__) +
+		             ": Exception in destructor",
+		             sinsp_logger::SEV_ERROR);
+	}
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_max_wait() const
+{
+	return m_max_wait_ms;
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_ttl() const
+{
+	return m_ttl_ms;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::stop()
+{
+	bool join_needed = false;
+
+	{
+		std::unique_lock<std::mutex> guard(m_mutex);
+
+		if(m_running)
+		{
+			m_terminate = true;
+			join_needed = true;
+
+			// The async thread might be waiting for new events
+			// so wake it up
+			m_queue_not_empty_condition.notify_one();
+		}
+	} // Drop the mutex before join()
+
+	if (join_needed)
+	{
+		m_thread.join();
+
+		// Remove any pointers from the thread to this object
+		// (just to be safe)
+		m_thread = std::thread();
+	}
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::is_running() const
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	return m_running;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::run()
+{
+	m_running = true;
+
+	while(!m_terminate)
+	{
+		{
+			std::unique_lock<std::mutex> guard(m_mutex);
+
+			while(!m_terminate && m_request_queue.empty())
+			{
+				// Wait for something to show up on the queue
+				m_queue_not_empty_condition.wait(guard);
+			}
+
+			prune_stale_requests();
+		}
+
+		if(!m_terminate)
+		{
+			run_impl();
+		}
+	}
+
+	m_running = false;
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::lookup(
+		const key_type& key,
+		metadata_type& metadata,
+		const callback_handler& callback)
+{
+	std::unique_lock<std::mutex> guard(m_mutex);
+
+	if(!m_running)
```

Comment:
as noted above, this needs to synchronize with the change of the m_running flag in the thread itself